syntax = "proto3";

package dispatcher;

message DeviceData {
    string id = 1;
    string mac_address = 2;
    float temperature = 3;
    float humidity = 4;
    float pressure = 5;
    float soil_moisture = 6;
    string timestamp = 7;
}

message Response {
    string status = 1;
    string message = 2;
    string data = 3;
}

message TimeRangeRequest {
    string mac_address = 1;
    string start_time = 2;
    string end_time = 3;
}

message AverageRequest {
    string mac_address = 1;
    string start_time = 2;
    string end_time = 3;
}

message AverageResponse {
    string status = 1;
    string message = 2;
    float avg_temperature = 3;
    float avg_humidity = 4;
}

message LastRecordRequest {
    string mac_address = 1;
}

message LastRecordResponse {
    string status = 1;
    string message = 2;
    DeviceData last_record = 3;
}

service Dispatcher {
    rpc SendDeviceData(DeviceData) returns (Response);
    rpc GetDataForPeriod(TimeRangeRequest) returns (Response);
    rpc GetAverageData(AverageRequest) returns (AverageResponse);
    rpc GetLastRecord(LastRecordRequest) returns (LastRecordResponse);
}
